Add runtime type-checking for `validate_call`

kc0506 opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->

## Change Summary

I added a more restrictive type-checking for input function, and give an informative error message in case of invalid inputs.

<!-- Please give a short summary of the changes. -->

## Related issue number

fix #10461
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

kc0506 commented:
I will add tests tomorrow 😃

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -11,89 +12,108 @@
 from . import _generate_schema, _typing_extra
 from ._config import ConfigWrapper
 
+_validate_call_handle_types: tuple[type[Any], ...] = (
```
Maybe supported types?

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -11,89 +12,108 @@
 from ._config import ConfigWrapper
 
+_validate_call_handle_types: tuple[type[Any], ...] = (
+    FunctionType,
```
Should we add `types.BuiltinFunctionType`?

sydney-runkle commented on pydantic/validate_call_decorator.py:
```diff
@@ -50,11 +50,7 @@ def validate_call(
-            raise TypeError(f'The `@{name}` decorator should be applied after `@validate_call` (put `@{name}` on top)')
-
-        validate_call_wrapper = _validate_call.ValidateCallWrapper(function, config, validate_return, local_ns)
+        validate_call_wrapper = _validate_call.wrap_validate_call(function, config, validate_return, local_ns)
```
I wonder if it makes sense to leave the function type checking logic in this file, just from a readability perspective.

Baiscally, check if `validate_call` supports wrapping said function, then define the validate function (see below)

sydney-runkle replied:
Side note - what's the case where `func` is None?

sydney-runkle replied:
Should we further abstract away some of this:

```py
if inspect.iscoroutinefunction(function):
    @functools.wraps(function)
    async def wrapper_function(*args, **kwargs):  # type: ignore
        return await validate_call_wrapper(*args, **kwargs)
else:
    @functools.wraps(function)
    def wrapper_function(*args, **kwargs):
        return validate_call_wrapper(*args, **kwargs)
wrapper_function.raw_function = function  # type: ignore
return wrapper_function  # type: ignore
```

into the `_validate_call.py` file?

kc0506 replied:
I think the type-checking should be inside `validate` since `func` could be `None` anyways. Maybe we could rename `validate` if it is for the sake of readability.

sydney-runkle replied:
> Side note - what's the case where func is None?

Do you know what the case is for this?

kc0506 replied:
This is for decorating without any args:
```py
@validate_call
def f(): ...
```

So in this we do not yet have the actual function inside `validate_call`, and we have to return a function which serves as a decorator.

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -11,89 +12,108 @@
+    namespace: dict[str, Any] | None,
+):
     """This is a wrapper around a function that validates the arguments passed to it, and optionally the return value."""
+    _check_function_type(function)
```
See comment below, would like to move this call to `validate_call.py`

sydney-runkle reviewed (commented):
Hi! Thanks for your contribution here! Looking forward to some tests.

Left a few comments.

kc0506 commented:
I've moved the supported types back to `_validate_call` because I we have to define some util functions (such as `get_qualname` currently), it will be more clear what possible type the input may be.

kc0506 commented:
I've added some tests. It is worth noting that since dunder methods cannot be overridden at instance level, it is impossible to validate a single callable instance; rather, the user have to validate the `__call__` at the class level, which affects all instances.

kc0506 commented on tests/test_validate_call.py:
```diff
@@ -26,19 +26,11 @@ def test_func_type():
-            with pytest.raises(TypeError, match=(f"Input function `{func.__name__}` doesn't have a valid signature")):
-                validate_call(func)
+        assert validate_call(func).__name__ == func.__name__
+        # validate_call(partial(func))
```
The commented line is currently failing because `_typing_extra.get_function_type_hints` does not work on built-in functions.

sydney-runkle commented:
Hey @kc0506, I presume you're ready for another review here?

kc0506 commented:
Yeah I think it's ready.

sydney-runkle commented:
Perfect, will take a look :).

kc0506 commented on tests/test_validate_call.py:
```diff
@@ -89,20 +206,6 @@ def foo_bar(a: int = None):
-    assert callable(foo_bar.raw_function)
-    assert repr(inspect.signature(foo_bar)) == '<Signature (a: int, b: int)>'
-
-
```
This is moved to the top.

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -11,89 +13,115 @@
 from . import _generate_schema, _typing_extra
 from ._config import ConfigWrapper
 
+# This should be aligned with `GenerateSchema.match_types`
+ValidateCallSupportedTypes = Union[
+    LambdaType,
+    FunctionType,
+    MethodType,
+    BuiltinFunctionType,
+    BuiltinMethodType,
+    functools.partial,
+]
```
Let's also define:

```python
validate_call_supported_types = get_args(ValidateCallSupportedTypes)
```

and use it in the relevant places in the code

Viicos reviewed (changes requested):
Not a full review, but first I don't think we should support built-in functions such as `sorted`.

The reason `inspect.signature` works on such built-ins is because it looks for the `__text_signature__` attribute. This is an implementation detail (CPython implements it for `sorted`, but not for `min`, `max` etc, while PyPy does).

Even if it works, such text signatures don't provide any type annotations, meaning Pydantic currently defaults to `Any`. This will result in confusing behavior:

```python
sorted.__text_signature__
#> '($module, iterable, /, *, key=None, reverse=False)'
validate_sorted = validate_call(sorted)

validate_sorted(some_iterable, reverse="not_a_bool")  # validates fine
```
Removing built-ins support will also allow you to switch back to `get_function_type_hints`, as fetching the annotation from the `Signature` object is going to cause a lot of issues (forward references etc).

kc0506 commented:
> Not a full review, but first I don't think we should support built-in functions such as sorted.

I am totally fine with not supporting built-in functions, as in most cases they can report errors when args is of wrong types. 

> Even if it works, such text signatures don't provide any type annotations, meaning Pydantic currently defaults to `Any`. This will result in confusing behavior:

My original thought on this is that we can add a new parameter `overrides` to provide custom signature/annotations:
```py
my_sorted = validate_call(overrides={'reverse': bool})(sorted)
```
But I am not certain how much this is better than just define a new function 🤔:
```
def my_sorted(arr: list[int], reverse: bool):
    return sorted(arr, reverse=reverse)
```

> 
> Removing built-ins support will also allow you to switch back to `get_function_type_hints`, as fetching the annotation from the `Signature` object is going to cause a lot of issues (forward references etc).

Just out of curiosity, can you show some example regarding this?

Viicos commented:
More importantly, validating built-in functions has little value. For almost all of them, Python implements checks at runtime (raising `TypeError`s in most cases).

> Just out of curiosity, can you show some example regarding this?

Haven't tried, but:

```python
Forward = int

@validate_call
def func(a: 'Forward'): pass
```

is probably going to fail.

kc0506 commented:
> Haven't tried, but:
> 
> ```python
> Forward = int
> 
> @validate_call
> def func(a: 'Forward'): pass
> ```
> 
> is probably going to fail.

Surprisingly, this actually works fine. The reason is that `_call_schema` will call `_generate_parameter_schema` on `'Forward'`, which is then passed to `_apply_annotations` and go through the whole process in `_generate_schema_inner`. 

Imo using annotations from `signature` seems more consistent and concise, if not causing any problem currently. Also, this way we are evaluating the annotation with an unified namespace (`GenerateSchema._types_namespace`), and we don't have to deal with wrapped functions such as `partial`, since `signature` should handle them well. 

> More importantly, validating built-in functions has little value. For almost all of them, Python implements checks at runtime (raising `TypeError`s in most cases).

Again, I totally agree with you on this. If we have to put a lot of effort only to support built-in functions, then we should just remove them from `ValidateCallSupportTypes` instead. But if everything works fine then there seems no reason to tell user not to validate them (despite being useless).

Viicos commented:
> Surprisingly, this actually works fine. The reason is that `_call_schema` will call `_generate_parameter_schema` on `'Forward'`, which is then passed to `_apply_annotations` and go through the whole process in `_generate_schema_inner`.

Ah yes, good catch, although this is not the case for `unpack_type = _typing_extra.unpack_type(annotation)`.

> But if everything works fine then there seems no reason to tell user not to validate them (despite being useless).

I still think it is going to be confusing. We could argue it is the same as having an unannotated function being decorated:

```python
@validate_call
def func(a): pass  # Why don't we explicitly disallow such use cases as well?
```

But users unfamiliar with the Python typing ecosystem might think that these functions are annotated, because type checkers use `typeshed` definitions. But on the other hand, the same can be said for stdlib functions (although the source code -- where functions are not annotated -- can still be accessed). Let's just make sure the `unpack_type` is resolved.

kc0506 commented:
> Ah yes, good catch, although this is not the case for `unpack_type = _typing_extra.unpack_type(annotation)`.

Added to test and fixed (I guess).

> But users unfamiliar with the Python typing ecosystem might think that these functions are annotated, because type checkers use `typeshed` definitions. But on the other hand, the same can be said for stdlib functions (although the source code -- where functions are not annotated -- can still be accessed). Let's just make sure the `unpack_type` is resolved.

I am convinced by this 👍

kc0506 commented:
Ah, now I get it. Perhaps we could add a section in the docs warning about the `typeshed` issue.

Viicos commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1897,7 +1901,11 @@ def _callable_schema(self, function: Callable[..., Any]) -> core_schema.CallSche
+                # Note: This was originally get by `_typing_extra.get_function_type_hints`,
+                #       but we switch to simply `p.annotation` to support bultins (e.g. `sorted`).
+                #       May need to revisit if anything breaks.
+                annotation = ForwardRef(p.annotation) if isinstance(p.annotation, str) else p.annotation
```
From the `_typing_extra` module:

```suggestion
                annotation = _make_forward_ref(p.annotation) if isinstance(p.annotation, str) else p.annotation
```

Viicos commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1897,7 +1901,11 @@ def _callable_schema(self, function: Callable[..., Any]) -> core_schema.CallSche
+                #       but we switch to simply `p.annotation` to support bultins (e.g. `sorted`).
+                #       May need to revisit if anything breaks.
+                annotation = ForwardRef(p.annotation) if isinstance(p.annotation, str) else p.annotation
+                annotation = _typing_extra.eval_type_backport(annotation, self._types_namespace)
```
```suggestion
                annotation = self._resolve_forward_ref(annotation)
```

(some places in `GenerateSchema` still use of `eval_type_backport` but let's not fix this here for now).

Viicos commented on pydantic/validate_call_decorator.py:
```diff
@@ -16,6 +17,50 @@
 
 
+def _check_function_type(function: object):
+    ERROR_CODE: PydanticErrorCodes = 'validate-call-type'
```
```suggestion
    ERROR_CODE = 'validate-call-type'
```

No need to explicitly annotate here

Viicos commented on docs/errors/usage_errors.md:
```diff
@@ -1238,3 +1238,105 @@ except PydanticUserError as exc_info:
+
+## Unsupported type for `validate_call` {#validate-call-type}
+
+`validate_call` has some limitations on the types it can validate. This error is raised when you try to use it with an unsupported type. Currently the supported types are:
+`LambdaType, FunctionType, MethodType, functools.partial`. For the case of `functools.partial`, the function being partially applied must be one of the supported types.
```
```suggestion
`validate_call` has some limitations on the callables it can validate. This error is raised when you try to use it with an unsupported callable. Currently the supported callables are functions (including lambdas), methods and instances of [`partial`][functools.partial]. In the case of [`partial`][functools.partial], the function being partially  applied must be one of the supported callables.
```

Viicos commented on docs/errors/usage_errors.md:
```diff
@@ -1238,3 +1238,105 @@ except PydanticUserError as exc_info:
+
+### `@classmethod`, `@staticmethod`, and `@property`
+
+These decorators cannot be put before `validate_call` because they return a class rather than one of the supported types. The correct way to use them is to put `validate_call` before the decorator.
```
```suggestion
These decorators must be put before `validate_call`.
```

I don't know if we should mention the fact that they are classes. It's not documented so could very much be a CPython implementation detail.

Viicos commented on docs/errors/usage_errors.md:
```diff
@@ -1238,3 +1238,105 @@ except PydanticUserError as exc_info:
+
+### Classes
+
+To avoid ambiguity, `validate_call` should be applied to methods, not classes. If you want to validate the constructor of a class, you should put `validate_call` on top of `__init__` or `__new__` instead.
```
```suggestion
While classes are callables themselves, `validate_call` can't be applied on them, as it needs to know about which method to use (`__init__` or `__new__`) to fetch type annotations. If you want to validate the constructor of a class, you should put `validate_call` on top of the appropriate method instead.
```

Viicos commented on docs/errors/usage_errors.md:
```diff
@@ -1238,3 +1238,105 @@ except PydanticUserError as exc_info:
+
+### Custom callable
+
+Although you can create custom callable types in Python with `__call__`, currently the instance of these types cannot be validated with `validate_call`. This may change in the future, but for now, you should use `validate_call` explicitly on `__call__` instead.
```
```suggestion
Although you can create custom callable types in Python by implementing a `__call__` method, currently the instances of these types cannot be validated with `validate_call`. This may change in the future, but for now, you should use `validate_call` explicitly on `__call__` instead.
```

Viicos reviewed (commented):
Definitely good to have these checks in place. Also great to have tests expanded!

However, I'm not sure about the removal of the `ValidateCallWrapper` class. I'm struggling to follow all the nested calls happening when applying the decorator and when calling the actual function for validation.

What benefits does the normal function pattern provide?

kc0506 commented:
The reason I consider `ValidateCallWrapper` unnecessary is that I've noticed it is not exposed anywhere. In the previous version, the only access to this class is through `__call__`, which means it basically served as a function, not a class instance. All the slots are unused by any other parts of the code.

So now I just return a `wrapper` function, which does the same thing as the original `__call__`. The whole logic is the same, just without an extra class.

Viicos commented:
Just chatted with @sydney-runkle, we think it's safer to keep the existing implementation with a class for now. The main reason is we might be able to expose it so that the `raw_function` method can be recognized from a type checking perspective. There have been a discussion about it [here](https://github.com/pydantic/pydantic/issues/9883#issuecomment-2222283570), where you'll find an example implementation.

Due to current typing limitations, we can't safely expose the class using `ParamSpec` without breaking type checking support for methods. But this may change in the future (see this [recent discussion](https://discuss.python.org/t/allow-self-binding-for-generic-paramspec/50948)), so that's why we feel it's safer to keep the existing implementation, which may change at some point.

Other than that, thanks for the effort put into details when it comes to tests and errors on invalid callables, definitely useful!

sydney-runkle commented:
@kc0506,

Following up on @Viicos' comment - thanks so much for your efforts here. We're super excited to adopt the new errors / docs / tests you've added here. We really appreciate your recent contributions and are excited to continue collaborating 🚀!

kc0506 commented:
> Due to current typing limitations, we can't safely expose the class using `ParamSpec` without breaking type checking support for methods. But this may change in the future (see this [recent discussion](https://discuss.python.org/t/allow-self-binding-for-generic-paramspec/50948)), so that's why we feel it's safer to keep the existing implementation, which may change at some point.
> 

Sure, do we want to keep the slots of `ValidateCallWrapper` then?

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -11,10 +13,56 @@
 class ValidateCallWrapper:
     """This is a wrapper around a function that validates the arguments passed to it, and optionally the return value."""
 
+    # This slots are not currently used, but in the future we may want to expose them.
+    # See #9883
```
It's true that having `__name__`/``__qualname__`/`__annotations__` as attributes on this class is weird, and confusing (because they are taken from the passed function, and they aren't the ones from this actual class). What happens if we remove them and we don't set them in `__init__`?

kc0506 replied:
As the comment says, nothing will happen because they are not used at all. I keeped it because if we return `ValidateCallWrapper` instead of a normal function in `validate_call`, then these slots would be necessary. However, removing them seems fine since we can just add them back without much effort and there is a test to ensure `__name__`, `__qualname__` etc. are correct.

Viicos replied:
Yes it may have been an oversight when this was implemented in V2. Let's leave them for now :)

kc0506 commented:
I made some refactors. Hope these do not make the reviewing harder 🙏.

I've:
1. extract `ValidateCallWrapper.__init__` into `_wrap_validate_call`
2. extract the logic of retieving function info (e.g. `__qualname__`) into seperate functions/classes.
3. simplify schema generation in `_wrap_validate_call`.

Viicos commented:
Thanks for continuing the work here. Can we please avoid refactoring things if it doesn't provide any benefit? :pray: It does make reviewing harder, makes it harder to git blame in the future, and from what I see it imo complicates things: `_get_wrap_info` is being called twice, the added level of indirection with added functions makes the logic harder to follow. Could you please revert the changes, and keep it as it was before? I think we can still keep the `update_wrapper` logic that should be called inside the `validate_call`'s nested `validate` function

sydney-runkle commented:
> Could you please revert the changes, and keep it as it was before?

@kc0506, once you've done this, I'm happy to take another pass :).

Viicos commented:
> once you've done this, I'm happy to take another pass :).

Feedback was already applied (but force pushed, that's why it doesn't appear in the last commit).

I've pushed a small change on `main` that simplify the `core_config` fetching logic, meaning you can remove the `str(...)` workaround ;)

sydney-runkle commented:
> I've pushed a small change on main that simplify the core_config fetching logic, meaning you can remove the str(...) workaround ;)

Let's do this, and @kc0506, I'll let you resolve merge conflicts 👍

kc0506 commented:
Conflicts have been resolved.

sydney-runkle commented on docs/errors/usage_errors.md:
```diff
@@ -1187,6 +1186,103 @@ except PydanticUserError as exc_info:
 
+## Unsupported type for `validate_call` {#validate-call-type}
+
+`validate_call` has some limitations on the callables it can validate. This error is raised when you try to use it with an unsupported callable. Currently the supported callables are functions (including lambdas), methods and instances of [`partial`][functools.partial]. In the case of [`partial`][functools.partial], the function being partially applied must be one of the supported callables.
```
```suggestion
`validate_call` has some limitations on the callables it can validate. This error is raised when you try to use it with an unsupported callable. Currently the supported callables are functions (including lambdas) and methods and instances of [`partial`][functools.partial]. In the case of [`partial`][functools.partial], the function being partially applied must be one of the supported callables.
```

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1837,16 +1837,17 @@ def _dataclass_schema(
                 return core_schema.definition_reference_schema(dataclass_ref)
 
-    def _callable_schema(self, function: Callable[..., Any]) -> core_schema.CallSchema:
+            # Type checkers seem to assume ExitStack may suppress exceptions and therefore
+            # control flow can exit the `with` block without returning.
+            assert False, 'Unreachable'
```
```suggestion
```

I think this is an artifact of a merge that can be removed...

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -12,10 +14,56 @@
 from ._config import ConfigWrapper
 from ._namespace_utils import MappingNamespace, NsResolver, ns_for_function
 
+# This should be aligned with `GenerateSchema.match_types`
+#
```
```suggestion
```

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -12,10 +14,56 @@
+    functools.partial,
+]
+
+VALIDATE_CALL_SUPPORTED_TYPES = get_args(ValidateCallSupportedTypes)
```
Why is this necessary? Can we not just use the above?

kc0506 replied:
At runtime, we might want to use `isinstance(obj, VALIDATE_CALL_SUPPORTED_TYPES)`

sydney-runkle commented on tests/test_validate_call.py:
```diff
@@ -1120,10 +1235,12 @@ def foo():
 
-        @validate_call
-        def bar(m: M) -> M:
+        @validate_call(validate_return=True)
+        def bar(m: 'M') -> 'M':
```
Why is this change necessary?

kc0506 replied:
To my understanding, we only use the namespaces to evaluate types when `str` is used to annotate things or `from __future__ import annotations` is used. Otherwise the annotations are normal Python objects and we do not evaluate them at all. Considering the purpose of this test, I would say the quotes are necessary to tell if namespaces are working correctly.

sydney-runkle commented on tests/test_validate_call.py:
```diff
@@ -347,32 +462,6 @@ def foo(args: int, kwargs: int):
     ]
 
 
-def test_v_args():
-    @validate_call
-    def foo1(v__args: int):
-        return v__args
-
-    assert foo1(123) == 123
-
-    @validate_call
-    def foo2(v__kwargs: int):
-        return v__kwargs
-
-    assert foo2(123) == 123
-
-    @validate_call
-    def foo3(v__positional_only: int):
-        return v__positional_only
-
-    assert foo3(123) == 123
-
-    @validate_call
-    def foo4(v__duplicate_kwargs: int):
-        return v__duplicate_kwargs
-
-    assert foo4(123) == 123
-
-
```
Why remove this?

kc0506 replied:
This looks like an incomplete/unnecessary test to me — the functions are nothing special and basially the same as simply
```py
@validate_call
def foo(x: int):
    return x
```
which obviously have been covered in other tests.

kc0506 replied:
It seems these are from `test_deprecated_validate_arguments.py`, where the argument names are invalid at the previous version. But in the new `validate_call` they are just normal names.

Viicos replied:
Yes can be removed after looking at the git blame.

sydney-runkle commented on pydantic/validate_call_decorator.py:
```diff
@@ -16,6 +17,50 @@
 
 
+def _check_function_type(function: object):
+    ERROR_CODE = 'validate-call-type'
```
Can be defined outside of the scope of the function, perhaps with a more descriptive name

sydney-runkle commented on pydantic/_internal/_validate_call.py:
```diff
@@ -12,10 +14,56 @@
+    return f'partial({func.func.__qualname__})' if isinstance(func, functools.partial) else func.__qualname__
+
+
+def update_wrapper(wrapped: ValidateCallSupportedTypes, wrapper: Callable[..., Any]):
```
Let's use a more descriptive name here

sydney-runkle commented on pydantic/_internal/_generate_schema.py:
```diff
@@ -1837,16 +1837,17 @@ def _dataclass_schema(
         """
         sig = signature(function)
 
-        globalns, localns = self._types_namespace
-        type_hints = _typing_extra.get_function_type_hints(function, globalns=globalns, localns=localns)
```
I'm concerned about the removal of this, cc @Viicos.

Viicos replied:
This is fine because the `self._resolve_forward_ref` will take care of it (and we used `ns_for_function` earlier on to get the ns)

sydney-runkle reviewed (changes requested):
Extra tests looking good for the most part, and the new docs are awesome. I've got some concerns about the changes in namespace management - I don't think we want to do that.

Thanks for your work here. Looking forward to including some of these improvements soon!

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -12,10 +14,56 @@
+def update_wrapper_attributes(wrapped: ValidateCallSupportedTypes, wrapper: Callable[..., Any]):
+    """Update the `wrapper` function with the attributes of the `wrapped` function. Return the updated function."""
+    if inspect.iscoroutinefunction(wrapped):
+        # We have to create a new couroutine function
```
```suggestion
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
 )
 
 
+def test_wrap():
```
```suggestion
def test_wrap() -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+    assert repr(inspect.signature(foo_bar)) == '<Signature (a: int, b: int)>'
+
+
+def test_func_type():
```
```suggestion
def test_func_type() -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+        validate_call([])
+
+
+def test_validate_class():
```
```suggestion
def test_validate_class() -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+    assert A('5').x == 5
+
+
+def test_validate_custom_callable():
```
```suggestion
def test_validate_custom_callable() -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+    assert B()('5') == 5
+
+
+def test_invalid_signature():
```
```suggestion
def test_invalid_signature() -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+
+
+@pytest.mark.parametrize('decorator', [staticmethod, classmethod])
+def test_classmethod_order_error(decorator):
```
```suggestion
def test_classmethod_order_error(decorator) -> None:
```

Viicos commented on tests/test_validate_call.py:
```diff
@@ -25,6 +25,130 @@
+                pass
+
+
 def test_args():
```
```suggestion
def test_args() -> None:
```

Viicos commented on pydantic/validate_call_decorator.py:
```diff
@@ -16,6 +17,55 @@
+_INVALID_TYPE_ERROR_CODE = 'validate-call-type'
+
+
+def _check_function_type(function: object):
```
```suggestion
def _check_function_type(function: object) -> None:
```

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -26,35 +74,35 @@ class ValidateCallWrapper:
         config: ConfigDict | None,
         validate_return: bool,
         parent_namespace: MappingNamespace | None,
     ):
```
```suggestion
    ) -> None:
```

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -26,35 +74,35 @@ class ValidateCallWrapper:
-            self.__module__ = function.__module__
+            module = function.__module__
+        qualname = core_config_title = extract_function_qualname(function)
+
+        self.__name__ = extract_function_name(function)
+        self.__qualname__ = qualname
+        self.__module__ = module
```
```suggestion
```

Viicos commented on pydantic/_internal/_validate_call.py:
```diff
@@ -12,10 +14,56 @@
 class ValidateCallWrapper:
     """This is a wrapper around a function that validates the arguments passed to it, and optionally the return value."""
 
+    # This slots are not currently used, but in the future we may want to expose them.
+    # See #9883
```
```suggestion
```

Viicos commented on docs/errors/usage_errors.md:
```diff
@@ -1065,7 +1065,7 @@ except PydanticUserError as exc_info:
 ## Cannot evaluate type annotation {#unevaluable-type-annotation}
 
-Because type annotations are evaluated *after* assignments, you might get unexpected results when using a type annotation name
+Because type annotations are evaluated _after_ assignments, you might get unexpected results when using a type annotation name
```
```suggestion
Because type annotations are evaluated *after* assignments, you might get unexpected results when using a type annotation name
```

Viicos commented:
Wrapping things up in https://github.com/pydantic/pydantic/pull/10627. Thanks @kc0506! Took some time but we got there.
